review bpf read user/kernel changes

hbrueckner opened the pull request:
review pr

hbrueckner committed:
new(bpf): enable helpers to read from kernel and user space memory

Similar to modern bpf, introduce a new argument that specifies from
where to read memory from.

Signed-off-by: Hendrik Brueckner <brueckner@de.ibm.com>

hbrueckner committed:
update(bpf): call helpers with user/kernel memory arguments

Review helpers and adjust them to read user or kernel memory depending
on the context.

Signed-off-by: Hendrik Brueckner <brueckner@de.ibm.com>

hbrueckner committed:
new(bpf): introduce bpf_probe_read_{kernel,user}* variants

Signed-off-by: Hendrik Brueckner <brueckner@linux.ibm.com>

hbrueckner committed:
new(bpf): provide _READP(P) variants for kernel/user memory reads

The BPF driver uses _READ(P) on mostly structures which reside in kernel
memory.  There are few exception where a _READ_USER(P) is required.

Introduce a _READ_KERNEL(P) and _READ_USER(P) variant and let the
_READ(P) always read from kernel memory (to keep the number of changes
at a minimum).

Signed-off-by: Hendrik Brueckner <brueckner@de.ibm.com>

hbrueckner committed:
update(bpf): correct places where _READ(P) reads from user memory

Signed-off-by: Hendrik Brueckner <brueckner@de.ibm.com>

hbrueckner committed:
update(bpf): correct bpf probe reads for kernel/user memory access

Review the `bpf_probe_read()` and `bpf_probe_read_str()` context and use
appropriate kernel/user variantes to read data.

Signed-off-by: Hendrik Brueckner <brueckner@de.ibm.com>

Molter73 commented on driver/bpf/filler_helpers.h:
```diff
@@ -859,17 +872,37 @@ static __always_inline int __bpf_val_to_ring(struct filler_data *data,
-						if (bpf_probe_read(&data->buf[curoff_bounded],
-								   ((read_size - 1) & SCRATCH_SIZE_HALF) + 1,
-								   (void *)val))
+						if (mem == KERNEL)
+						{
+							rc = bpf_probe_read_kernel(&data->buf[curoff_bounded],
+									  ((read_size - 1) & SCRATCH_SIZE_HALF) + 1,
+									  (void *)val);
+						}
+						else
+						{
+							rc = bpf_probe_read_user(&data->buf[curoff_bounded],
+									  ((read_size - 1) & SCRATCH_SIZE_HALF) + 1,
+									  (void *)val);
+						}
```
Can we put a couple of curly braces around this block of code? It would make it easier to see the entire block is executed only is `read_size` is true.

Molter73 replied:
Also, maybe we can come up with a macro or function that would make this check internally, I see the same check is done multiple times and makes it a bit harder to follow.

hbrueckner replied:
@Molter73 I extracted a helper routine.

Molter73 commented on driver/bpf/filler_helpers.h:
```diff
@@ -892,14 +926,33 @@ static __always_inline int __bpf_val_to_ring(struct filler_data *data,
-					if (bpf_probe_read(&data->buf[curoff_bounded],
-							   ((read_size - 1) & SCRATCH_SIZE_HALF) + 1,
-							   (void *)val))
+					if (mem == KERNEL)
+					{
+						rc = bpf_probe_read_kernel(&data->buf[curoff_bounded],
+								  ((read_size - 1) & SCRATCH_SIZE_HALF) + 1,
+								  (void *)val);
+					}
+					else
+					{
+						rc = bpf_probe_read_user(&data->buf[curoff_bounded],
+								  ((read_size - 1) & SCRATCH_SIZE_HALF) + 1,
+								  (void *)val);
+					}
```
Same as my 2 previous comments, put curly braces around the block or move it into its own macro/function.

Molter73 commented on driver/bpf/filler_helpers.h:
```diff
@@ -994,6 +1047,51 @@ static __always_inline int __bpf_val_to_ring(struct filler_data *data,
+	case PT_FSPATH:
+	case PT_FSRELPATH:
+	case PT_BYTEBUF:
+	{
+		/* Those above require some review on whether USER reads are
+		 * correct. If not, explicit use the respective helper with
+		 * the _mem() suffix to specify the memory to read from.
+		 *
+		 * See also the usage below in the helpers.
+		 */
+		return USER;
+	}
+	default:
+	{
+		return KERNEL;
+	}
```
[nit] I don't think you need braces here
```suggestion
		/* Those above require some review on whether USER reads are
		 * correct. If not, explicit use the respective helper with
		 * the _mem() suffix to specify the memory to read from.
		 *
		 * See also the usage below in the helpers.
		 */
		return USER;

	default:
		return KERNEL;
```

hbrueckner replied:
Done. Thank you!

Molter73 commented on driver/bpf/filler_helpers.h:
```diff
@@ -25,6 +25,15 @@ or GPL2.txt for full copies of the license.
+ */
+enum read_memory
+{
+	USER = 0,
+	KERNEL = 1,
```
This is an ultra-nitpick, but after seeing these enums in the code it is a little cryptic what `USER` and `KERNEL` might refer to, maybe we can rename them to something like `USER_MEMORY` and `KERNEL_MEMORY`, WDYT?

hbrueckner replied:
I actually simply re-used `enum read_memory` from the modern bpf driver. Let me check try to find a place to have it common to both bpf drivers. We can then have a follow-up change for all places in bpf and moden bpf driver.

Molter73 committed:
Fix eBPF compilation on kernels <= 5.5

hbrueckner committed:
update(bpf): add feature gate for `bpf_probe_read_{kernel|user}*()` variants

The `bpf_probe_read_{kernel|user}*()` functions are available on recent
kernel versions only.  Consolidate the version check as feature gate.

Signed-off-by: Hendrik Brueckner <brueckner@de.ibm.com>

hbrueckner committed:
update(bpf): clean-up curly braces

Signed-off-by: Hendrik Brueckner <brueckner@de.ibm.com>
Co-authored-by: Mauro Ezequiel Moltrasio <mmoltras@redhat.com>

hbrueckner committed:
update(bpf): Extract helper for read values into the ring

Signed-off-by: Hendrik Brueckner <brueckner@de.ibm.com>
Co-authored-by: Mauro Ezequiel Moltrasio <mmoltras@redhat.com>

hbrueckner committed:
update(bpf): add RHEL version check for USE_BPF_PROBE_KERNEL_USER_VARIANTS

Signed-off-by: Hendrik Brueckner <brueckner@de.ibm.com>
Co-authored-by: Mauro Ezequiel Moltrasio <mmoltras@redhat.com>

hbrueckner committed:
update(bpf): remove duplicate _READ in bpf_addr_to_kernel

Callers specify the len directly instead of a pointer to read it.

Signed-off-by: Hendrik Brueckner <brueckner@de.ibm.com>

hbrueckner committed:
fixup! update(bpf): correct bpf probe reads for kernel/user memory access

correctly read arguments from task mm struct (it is user space memory)

Signed-off-by: Hendrik Brueckner <brueckner@de.ibm.com>

Molter73 commented on driver/bpf/filler_helpers.h:
```diff
@@ -859,17 +893,15 @@ static __always_inline int __bpf_val_to_ring(struct filler_data *data,
 
 				if (!data->curarg_already_on_frame) {
 					volatile u16 read_size = dpi_lookahead_size;
+					int rc;
```
I just realized that when `BPF_FORBIDS_ZERO_ACCESS` is defined and `read_size` is 0, `rc` is not set, should we initialize it?
```suggestion
					int rc = 0;
```
